keystorage/src/main/java/tech/pegasys/web3signer/keystorage/awskms/AwsKeyManagerService.java

```diff
@@ -0,0 +1,172 @@
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.kms.KmsClient;
+import software.amazon.awssdk.services.kms.KmsClientBuilder;
+import software.amazon.awssdk.services.kms.model.GetPublicKeyRequest;
+import software.amazon.awssdk.services.kms.model.GetPublicKeyResponse;
+import software.amazon.awssdk.services.kms.model.KeySpec;
+import software.amazon.awssdk.services.kms.model.MessageType;
+import software.amazon.awssdk.services.kms.model.SignRequest;
+import software.amazon.awssdk.services.kms.model.SignResponse;
+import software.amazon.awssdk.services.kms.model.SigningAlgorithmSpec;
+
+/**
+ * Wraps AWS KMS Client and expose public key and sign operations. Since KmsClient is thread safe,
+ * same instance of this class can be used to talk to same host/credentials.
+ */
+public class AwsKeyManagerService implements AutoCloseable {
+  private final AwsAuthenticationMode authMode;
+  private final AwsCredentials awsCredentials;
+  private final String region;
+
+  private KmsClient kmsClient;
+
+  private Optional<URI> endpointOverride;
+
+  public AwsKeyManagerService(
+      final AwsAuthenticationMode authMode,
+      final AwsCredentials awsCredentials,
+      final String region,
+      final Optional<URI> endpointOverride) {
+    this.authMode = authMode;
+    this.awsCredentials = awsCredentials;
+    this.region = region;
+    this.endpointOverride = endpointOverride;
+
+    initKmsClient();
+  }
+
+  void initKmsClient() {
+    if (kmsClient != null) {
+      return;
+    }
+
+    final KmsClientBuilder kmsClientBuilder = KmsClient.builder();
+    endpointOverride.ifPresent(kmsClientBuilder::endpointOverride);
+    kmsClient =
+        kmsClientBuilder
+            .credentialsProvider(getAwsCredentialsProvider())
+            .region(Region.of(region))
+            .build();
+  }
+
+  private AwsCredentialsProvider getAwsCredentialsProvider() {
+    final AwsCredentialsProvider awsCredentialsProvider;
+    if (authMode == ENVIRONMENT) {
+      awsCredentialsProvider = DefaultCredentialsProvider.create();
+    } else {
+      final software.amazon.awssdk.auth.credentials.AwsCredentials awsSdkCredentials;
+      if (awsCredentials.getSessionToken().isPresent()) {
+        awsSdkCredentials =
+            AwsSessionCredentials.create(
+                awsCredentials.getAccessKeyId(),
+                awsCredentials.getSecretAccessKey(),
+                awsCredentials.getSessionToken().get());
+      } else {
+        awsSdkCredentials =
+            AwsBasicCredentials.create(
+                awsCredentials.getAccessKeyId(), awsCredentials.getSecretAccessKey());
+      }
+
+      awsCredentialsProvider = StaticCredentialsProvider.create(awsSdkCredentials);
+    }
+    return awsCredentialsProvider;
+  }
+
+  public ECPublicKey getECPublicKey(final String kmsKeyId) {
+    checkArgument(kmsClient != null, "KmsClient is not initialized");
+
+    // Question ... do we need to set grantTokens?
```

Comment:
Question for yourself or reviewers? :)

Reply:
for both :). The API talks about using `grantTokens` to further restrict the authentication ... however, I am not sure whether we want to support it or not. May be we can take it as optional parameters from our meta file?

Reply:
Is this a more general AWS feature, i.e. could be used with AWSSecretsManager for eth2 as well? If so, then I'd say it should be split out to another issue, especially if it isn't blocking using eth1 with AWS auth in the same way as we already use it for eth2.